hax/hax/motr/planner.py

```diff
@@ -0,0 +1,297 @@
+# Copyright (c) 2020 Seagate Technology LLC and/or its Affiliates
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+#
+
+import logging
+from collections import deque
+from dataclasses import dataclass
+from threading import Condition
+from typing import Callable, Deque, Optional, Set, Type
+
+from hax.log import TRACE
+from hax.message import (AnyEntrypointRequest, BaseMessage, BroadcastHAStates,
+                         Die, HaNvecGetEvent, SnsOperation)
+from hax.motr.util import LinkedList
+
+LOG = logging.getLogger('hax')
+MAX_GROUP_ID = 100000
+
+__all__ = ['WorkPlanner']
+
+
+@dataclass
+class State:
+    #
+    # group being executed currently
+    current_group_id: int
+    #
+    # group that is being populated by next add_command() invocation
+    next_group_id: int
+    #
+    # commands that are being processed now
+    active_commands: LinkedList[BaseMessage]
+    #
+    # The commands that have already been taken by the worker threads
+    # but not approved yet for execution (so either threads haven't invoked
+    # ensure_allowed() yet or ensure_allowed() has not succeded for them).
```

Comment:
So what happens to the commands for whom ensure_allowed() fails? do they get assigned to a different group then?

Reply:
There is no definition of 'fails' in our code.

From WorkPlanner's point of view, the command can be planned for processing ('taken'), it can be in progress now (being processed) and finished.

If an exception happens while processing a command, it should be handled within ConsumerThread (and [it is done](https://github.com/Seagate/cortx-hare/blob/3ff7c323438c67772dd433505986dfe70a44996f/hax/hax/handler.py#L62-L63) [like that](https://github.com/Seagate/cortx-hare/blob/3ff7c323438c67772dd433505986dfe70a44996f/hax/hax/handler.py#L224-L230) already). It is up to the ConsumerThread logic to decide whether the operation should be retried or just skipped. If the operation is retried, we do not leverage any special WorkPlanner mechanisms. All the retries (e.g. because of Consul issues) are looking as one long processing of the same command.